src/puppeteer_utils.js

```diff
@@ -348,6 +345,58 @@ const compileScript = (scriptString, context = Object.create(null)) => {
     return func;
 };
 
+/**
+ * Extended version of page.goto() allowing to perform requests with HTTP method other than GET,
+ * with custom headers and POST payload. URL, method, headers and payload are taken from
+ * request parameter that must be an instance of Apify.Request class.
+ *
+ * @param {Page} page
+ *   Puppeteer <a href="https://pptr.dev/#?product=Puppeteer&show=api-class-page" target="_blank"><code>Page</code></a> object.
+ * @param {Request} request {@link Request} object.
+ * @param {Object} gotoOptions Custom options for page.goto().
+ * @return {Promise<Response>}
+ *
+ * @memberOf puppeteer
+ */
+export const gotoExtended = async (page, request, gotoOptions = {}) => {
+    checkParamOrThrow(page, 'page', 'Object');
+    checkParamPrototypeOrThrow(request, 'request', Request, 'Apify.Request');
+    checkParamOrThrow(gotoOptions, 'gotoOptions', 'Object');
+
+    const { method, headers, payload } = request;
+
+    if (method !== 'GET' || payload || !_.isEmpty(headers)) {
+        const interceptRequestHandler = async (interceptedRequest) => {
+            const overrides = {};
+
+            if (method !== 'GET') overrides.method = method;
+            if (payload) overrides.postData = payload;
+            if (!_.isEmpty(headers)) overrides.headers = headers;
```

Comment:
I thought we were going to merge the headers?

Reply:
We could change this in `handleRequest()` of `src/puppeteers_request_interception.js` which would be a bit cleaner by changing

```js
     const requestOverrides = {};

     const originalContinue = request.continue.bind(request);
     request.continue = (overrides) => {
         wasContinued = true;
         Object.assign(requestOverrides, overrides);
     };
```

to:

```js
     const requestOverrides = {
         headers: request.headers(),
     };

     const originalContinue = request.continue.bind(request);
     request.continue = (overrides) => {
         wasContinued = true;
         const headers = Object.assign({}, requestOverrides.headers, overrides.headers);
         Object.assign(requestOverrides, overrides, { headers });
     };
```

Reply:
Yeah, that is a good idea. Although I'd probably rename the `requestOverrides`.

What about `finalOverrides` or `accumulatedOverrides`?